### ISSUE TITLE
Proposal: ScriptTool with customizable interpolation delimiters

### ISSUE THREAD

**Issue by TMiguelT** [common-workflow-language/common-workflow-language#852]
*Created with consultation with @kaushik-work*

# Use cases

It is a common desire to run an entire script as a CWL step. For example, the user might be converting a `bash` or `make`-based pipeline to CWL and would like to convert entire logical blocks of shell code into CWL.

The current approaches for doing this, either via the `InitialWorkDirRequirement` ([e.g. this example from the user guide](https://www.commonwl.org/user_guide/14-runtime/index.html)), or via `valueFrom` ([e.g. this example](https://github.com/EBI-Metagenomics/ebi-metagenomics-cwl/blob/master/tools/fasta_chunker.cwl#L25)) this has two major flaws:
 * The syntax is awkward and verbose. It is unclear why a script is part of the working directory requirement in the first example, and unclear why a script is the value for a command line parameter in the second example. The syntax for expressing this is fairly complex in both cases
* CWL's `$()` interpolation can interfere with the variable interpolation of other languages, via the use of the dollar sign, which is a common syntax element in other languages

# Proposed solution
We add a new `Process` object called `ScriptTool`, similar to `ExpressionTool` and `CommandLineTool`, which describes an interpreted script, written in any language. The new type requires an interpreter, the literal script, and optionally defines delimiters for variable interpolation. We also allow common requirements such as `DockerRequirement`, `ResourceRequirement` etc. `ScriptTool` has otherwise the same fields as a `CommandLineTool` in terms of inputs and outputs.

# Example Syntax
Here I have converted the bash script example described in the User Guide [here](https://www.commonwl.org/user_guide/14-runtime/index.html), to use my new syntax. 

Note that we are able to use the standard Bash variable interpolation because we have defined a custom delimiter.
```
class: ScriptTool
cwlVersion: v1.0
interpreter: bash

delims: [~<, >]

script:
  PREFIX='Message is:'
  MSG="${PREFIX} ~<inputs.message>"
  echo ${MSG}

inputs:
  message: string
outputs:
  example_out:
    type: stdout
stdout: output.txt
```

# Fields
* `delims` is an array of exactly two strings. The first entry is the string indicating the start of a variable interpolation, and the second is indicates the end of interpolation
* `script` is a string type (*not* an `Expression`), which will optionally be interpolated, before it is saved to a file and run by the interpreter
* `interpreter` is a string type, which defines the path (absolute, or within $PATH) to the interpreter that will interpret this script.
* `inputs` is a list of inputs, which may not have the type `CommandInputParameter`, because they shouldn't have an input binding. This is because the inputs will only ever be interpolated into the script, they will never be used on the command line

# Discussion

* "What if we I need multiple scripts that call each other?" - The secondary scripts can be defined in the Docker image, or using `InitialWorkDirRequirement`. The `ScriptTool` then becomes the entry point script, which can call other scripts. Doing it this way saves having a more complex syntax that supports multiple scripts with multiple interpreters and delimiters.
* "What if I want to disable variable interpolation?" - Omitting the `delims` field means disabling variable interpolation
* "Can't we already do this with either `InitialWorkDirRequirement` or `valueFrom`"? Yes, but in CWL we try to be explicit about the purpose of each tool. Neither of these options makes it clear that our main unit of execution is in fact a script. In addition, this proposal allows customizable variable interpolation
* "Doesn't this complicate the CWL spec, if there are existing ways to do this?" - yes, but this lowers the cognitive burden on those reading and writing CWL. It means they only need to learn the simple `ScriptTool` syntax if they are converting a script to CWL, instead of the complexity of `CommandLineTool`. Again, being concise is ideal

# Origins
This proposal is inspired by some of the discussion here: https://github.com/common-workflow-language/common-workflow-language/issues/850

**Comment by tetron** [comment]
Here's my take on customized delimiters (arbitrary delimiters isn't necessary, just the prefix) https://github.com/common-workflow-language/cwltool/pull/1132

**Comment by KerstenBreuer** [comment]
I agree to @TMiguelT. Please take this serious. In bioinformatics, nearly everything is happening in scripts and command line tools are rather an exception. To me a ScriptTool would be essential. One of the reasons why snakemake is so successful (at least in germany) is that you can easily copy in your script and specify the interpreter. The same experience is required for CWL.
